sui_programmability/examples/nfts/sources/random_word.move

```diff
@@ -0,0 +1,150 @@
+
+    /// Create a new word NFTs shop for a given set of words and their weights.
+    public entry fun create(words: vector<String>, weights: vector<u8>, ctx: &mut TxContext) {
+        assert!(vector::length(&words) == vector::length(&weights), EInvalidSize);
+        let total_weight: u64 = 0;
+        let i = 0;
+        while (i < vector::length(&weights)) {
+            total_weight = total_weight + (*vector::borrow(&weights, i) as u64);
+            i = i + 1;
+        };
+
+        let shop = Shop {
+            id: object::new(ctx),
+            issuer: tx_context::sender(ctx),
+            words,
+            weights,
+            total_weight,
+        };
+        sui::transfer::freeze_object(shop);
+    }
+
+    /// Buy a random word:
+    /// - Pay 1 SUI to to issuer.
+    /// - Create a randomness object.
+    /// - Create a ticket that references the randomness object.
+    /// - Transfer both objects to the caller.
+    public entry fun buy(shop: &Shop, coin: Coin<SUI>, ctx: &mut TxContext) {
+        assert!(coin::value(&coin) == 1, 0);
+        sui::transfer::transfer(coin, shop.issuer);
+
+        let randomness = randomness::new(RANDOMNESS_WITNESS {}, ctx);
+        let ticket = Ticket {
+            id: object::new(ctx),
+            shop_id: object::id(shop),
+            randomness_id: object::id(&randomness),
+        };
+        randomness::transfer(randomness, tx_context::sender(ctx));
+        sui::transfer::transfer(ticket, tx_context::sender(ctx));
+    }
+
+    /// Set the Randomness object with the given signature, and create the derived word NFT.
+    public entry fun mint(
+        shop: &Shop,
+        ticket: Ticket,
+        randomness: randomness::Randomness<RANDOMNESS_WITNESS>,
+        sig: vector<u8>,
+        ctx: &mut TxContext
+    ) {
+        assert!(ticket.shop_id == object::id(shop), EWrongShop);
+        assert!(ticket.randomness_id == object::id(&randomness), EWrongRandomness);
+
+        randomness::set(&mut randomness, sig);
+        let random_value = option::borrow(randomness::value(&randomness));
+        let selection = safe_selection(shop.total_weight, random_value);
+
+        // Find the relevant word by iterating over the weights.
+        let curr_weight: u64 = 0;
+        let i = 0;
+        while (curr_weight < shop.total_weight) {
+            let weight = (*vector::borrow(&shop.weights, i) as u64);
+            if (selection < curr_weight + weight) {
+                let word = Word {
+                    id: object::new(ctx),
+                    word: *vector::borrow(&shop.words, i),
+                    creator: object::id(shop),
+                };
+                sui::transfer::transfer(word, tx_context::sender(ctx));
+                break
+            };
+            curr_weight = curr_weight + weight;
+            i = i + 1;
+        };
+
+        // Release objects.
+        randomness::destroy(randomness);
+        destroy_ticket(ticket);
```

Comment:
Nit: If that's the only place `destroy_ticket` is used, it might be better to do destruction here. Mostly for gas reasons - function calls are more expensive.